File: python/semantic_kernel/utils/telemetry/user_agent.py
# Copyright (c) Microsoft. All rights reserved.

import os
from importlib.metadata import PackageNotFoundError, version
from typing import Any

from semantic_kernel.const import USER_AGENT

# Note that if this environment variable does not exist, telemetry is enabled.
TELEMETRY_DISABLED_ENV_VAR = "AZURE_TELEMETRY_DISABLED"

IS_TELEMETRY_ENABLED = os.environ.get(TELEMETRY_DISABLED_ENV_VAR, "false").lower() not in ["true", "1"]

HTTP_USER_AGENT = "semantic-kernel-python"

try:
    version_info = version("semantic-kernel")
except PackageNotFoundError:
    version_info = "dev"

APP_INFO = (
    {
        "semantic-kernel-version": f"python/{version_info}",
    }
    if IS_TELEMETRY_ENABLED
    else None
)


SEMANTIC_KERNEL_USER_AGENT = f"{HTTP_USER_AGENT}/{version_info}"


def prepend_semantic_kernel_to_user_agent(headers: dict[str, Any]):
    """Prepend "semantic-kernel" to the User-Agent in the headers.

    Args:
        headers: The existing headers dictionary.

    Returns:
        The modified headers dictionary with "semantic-kernel-python/{version}" prepended to the User-Agent.
    """
    headers[USER_AGENT] = (
        f"{SEMANTIC_KERNEL_USER_AGENT} {headers[USER_AGENT]}" if USER_AGENT in headers else SEMANTIC_KERNEL_USER_AGENT
    )

    return headers

File: nodejs/semantic_kernel/utils/telemetry/user-agent.ts
import * as packageJson from '../../../package.json'

const USER_AGENT = 'User-Agent'

// Note that if this environment variable does not exist, telemetry is enabled.
const TELEMETRY_DISABLED_ENV_VAR = 'AZURE_TELEMETRY_DISABLED'

export const IS_TELEMETRY_ENABLED =
  (process.env[TELEMETRY_DISABLED_ENV_VAR] || 'false').toLowerCase() !== 'true' &&
  (process.env[TELEMETRY_DISABLED_ENV_VAR] || 'false').toLowerCase() !== '1'

export const HTTP_USER_AGENT = 'semantic-kernel-nodejs'

let versionInfo

try {
  // Try to load version from package.json
  // In a bundled environment, this may not work, so we use 'dev' as fallback
  versionInfo = packageJson.version || 'dev'
} catch {
  // If package.json cannot be loaded, use 'dev'
  versionInfo = 'dev'
}

export const APP_INFO = IS_TELEMETRY_ENABLED
  ? {
      'semantic-kernel-version': `nodejs/${versionInfo}`,
    }
  : null

export const SEMANTIC_KERNEL_USER_AGENT = `${HTTP_USER_AGENT}/${versionInfo}`

/**
 * Prepend "semantic-kernel" to the User-Agent in the headers.
 *
 * @param headers - The existing headers dictionary
 * @returns The modified headers dictionary with "semantic-kernel-nodejs/{version}" prepended to the User-Agent
 */
export function prependSemanticKernelToUserAgent(headers: Record<string, any>): Record<string, any> {
  headers[USER_AGENT] = headers[USER_AGENT]
    ? `${SEMANTIC_KERNEL_USER_AGENT} ${headers[USER_AGENT]}`
    : SEMANTIC_KERNEL_USER_AGENT

  return headers
}
